### ISSUE TITLE
Tests fail with the latest dmd version (2.071.0)

### ISSUE THREAD

**Issue by BlackEdder** [BlackEdder/painlessjson#52]
This is because isFieldOrProperty!(Tuple.toString) returns true:
https://issues.dlang.org/show_bug.cgi?id=15899

Is it worth it to temporary work around this by either accepting (in the test) that Tuple.toJSON adds a toString field, or test specifically for toString in isFieldOrProperty. Alternatively we wait for the bug to be fixed.

**Comment by Zalastax** [comment]
I did some research and couldn't find any clean solutions. The trait isFieldOrProperty should preferably find out if T actually is a field, but I don't know how. A specific test in isFieldOrProperty is my bet.

**Comment by BlackEdder** [comment]
Are there other options beside field or function? I thought !isSomeFunction would therefore mean it is a field.

**Comment by BlackEdder** [comment]
Never mind, apparently template is a third option.

To check whether something is a field we could see if it is in .tupleof. According to the documentation that will only returns fields (including hidden fields though).

We could actually just call tupleof and then convert the resulting tuple to json.

**Comment by Zalastax** [comment]
That sounds like a good start. `excluding the hidden fields and the fields in the base class.`; Not getting fields from the base class clashes with the current implementation, right? https://dlang.org/phobos/std_traits.html#Fields is interesting as well. .tupleof / Fields doesn't return member functions so property functions has to be found in another way if they are to be used.

**Comment by BlackEdder** [comment]
Fields sounds indeed like what we would want. For the properties we can still use if isSomeFunction, see if it is a property.

You could actually argue that we shouldn't include properties by default (make it a config switch), but I'd rather not change painlessjson's default behaviour.

**Comment by BlackEdder** [comment]
Fields actually also doesn't work, because it does not include the names of the fields. We could use FieldNameTuple instead. Doing that will require a bit of work though. Instead I went for the simpler solution of explicitly checking whether something is a template in isFieldOrProperty (see msoucy/painlesstraits#4). I also suggested adding an isField function in painlesstraits that returns true if something is not a function/delegate or template.

This issue can be closed when the changes are merged msoucy/painlesstraits#4 and version number is updated.

EDIT: clarity and mention FieldNameTuple
